File: ihchat/php/app/Atendimento/Distribuicao.php
<?php
declare(strict_types=1);

namespace IHchat\Atendimento;

use IHchat\Auth\Atendentes;
use IHchat\Banco\Banco;

/** Distribuição de conversas entre os atendentes (app/servicos/distribuicao.py). */
final class Distribuicao
{
    /**
     * Menor fila primeiro: entrega a quem tem menos conversas em andamento.
     * Empate desfeito pelo id, o que faz a distribuição circular entre
     * atendentes com a mesma carga.
     *
     * @return array<string, mixed>|null o atendente (tipado) ou null se ninguém está disponível
     */
    public static function proximoAtendente(): ?array
    {
        $linha = Banco::um(
            "SELECT a.* FROM atendentes a
             LEFT JOIN (SELECT atendente_id, COUNT(id) AS total FROM conversas
                        WHERE status <> 'resolvida' AND atendente_id IS NOT NULL
                        GROUP BY atendente_id) carga ON carga.atendente_id = a.id
             WHERE a.ativo = 1 AND a.disponivel = 1
             ORDER BY COALESCE(carga.total, 0) ASC, a.id ASC
             LIMIT 1"
        );
        return $linha === null ? null : Atendentes::tipar($linha);
    }
}

File: ihchat/app/servicos/distribuicao.py
"""Distribuicao de conversas entre os atendentes."""
from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from ..models import Atendente, Conversa, StatusConversa


def proximo_atendente(sessao: Session, setor_id: int | None = None) -> Atendente | None:
    """Menor fila primeiro: entrega a quem tem menos conversas em andamento.

    Empate e desfeito pelo id, o que faz a distribuicao circular entre
    atendentes com a mesma carga. Conversa de um setor só vai para quem está
    disponível NAQUELE setor; sem ninguém, fica sem atendente na fila do
    setor (não "vaza" para outro). Conversa sem setor vai para qualquer um.
    """
    carga = (
        select(Conversa.atendente_id, func.count(Conversa.id).label("total"))
        .where(Conversa.status != StatusConversa.RESOLVIDA.value)
        .group_by(Conversa.atendente_id)
        .subquery()
    )
    consulta = (
        select(Atendente)
        .outerjoin(carga, carga.c.atendente_id == Atendente.id)
        .where(Atendente.ativo.is_(True), Atendente.disponivel.is_(True))
        .order_by(func.coalesce(carga.c.total, 0).asc(), Atendente.id.asc())
        .limit(1)
    )
    if setor_id is not None:
        consulta = consulta.where(Atendente.setor_id == setor_id)
    return sessao.scalar(consulta)
